fix: stats column binary_column has unsupported type binary

omkar-foss opened the pull request:
# Description
Handle `BINARY` column appropriately when specified in `delta.dataSkippingStatsColumns` while writing to a delta table.

# Related Issue(s)
<!---
For example:

- closes #106
--->

- closes https://github.com/delta-io/delta-rs/issues/2854
- related https://github.com/delta-io/delta-rs/issues/2373

# Related PR(s)

- https://github.com/delta-io/delta-rs/pull/1766
- https://github.com/delta-io/delta-rs/pull/1498 (reverted in PR https://github.com/delta-io/delta-rs/pull/1544)
- 

# Documentation

<!---
Share links to useful documentation
--->

N/A

ion-elgreco commented:
@omkar-foss are you still working on a fix :)?

omkar-foss commented:
I checked the delta lake protocol, and it looks like the stats are processed only for numeric columns types and stored in this structure: https://github.com/delta-io/delta/blob/master/PROTOCOL.md?plain=1#L1942-L1955

Since minValues and maxValues cannot be computed for BINARY columns, one option would be that we keep BINARY columns excluded from column statistics altogether.

Another option could be that we support BINARY columns and only compute numRecords and nullCount for them. minValues and maxValues will be unavailable.

@ion-elgreco any thoughts on which way we should go? My inclination is towards keeping them excluded.

ion-elgreco commented:
@omkar-foss let's exclude it then yeah, maybe add a warm trace when such a column was set

omkar-foss commented:
> @omkar-foss let's exclude it then yeah, maybe add a warm trace when such a column was set

Yes sure, I'll update this PR accordingly. Thank you :)

omkar-foss commented:
> > @omkar-foss let's exclude it then yeah, maybe add a warm trace when such a column was set
> 
> Yes sure, I'll update this PR accordingly. Thank you :)

@ion-elgreco I've updated the PR, also added a few tests on python and rust side to cover the fix. Let me know if any further changes or improvements, thank you :)

ion-elgreco commented on crates/core/src/kernel/snapshot/mod.rs:
```diff
@@ -652,9 +653,20 @@ fn stats_schema(schema: &StructType, config: TableConfig<'_>) -> DeltaResult<Str
     let stats_fields = if let Some(stats_cols) = config.stats_columns() {
         stats_cols
             .iter()
+            .filter(|col| {
```
Maybe use filter_map here

omkar-foss replied:
Done, thanks!

ion-elgreco commented on python/src/lib.rs:
```diff
@@ -390,7 +391,22 @@ impl RawDeltaTable {
                 .config()
                 .stats_columns()
-                .map(|v| v.iter().map(|s| s.to_string()).collect::<Vec<String>>()))
+                .map(|v| v.iter().map(|s| s.to_string())
+                .filter(| s | {
```
We shouldnt filter here actually

omkar-foss replied:
It looks like if we don't filter here, then the result of `get_stats_columns()` returns the binary column as well. Without a filter here, [this test](https://github.com/omkar-foss/delta-rs/blob/fix-dssc-binary-col/python/tests/test_writer.py#L1952-L1974) fails with the following error:
```bash
>       assert dt._table.get_stats_columns() == ["y"]
E       AssertionError: assert ['x', 'y'] == ['y']
E         
E         At index 0 diff: 'x' != 'y'
E         Left contains one more item: 'y'
E         
E         Full diff:
E           [
E         +     'x',
E               'y',
E           ]

tests/test_writer.py:1974: AssertionError
```

ion-elgreco commented on crates/core/src/operations/write/mod.rs:
```diff
@@ -599,9 +599,33 @@ impl std::future::IntoFuture for WriteBuilder {
 
+            stats_columns = match stats_columns {
+                Some(columns) => columns.iter().filter(| s | {
+                    let (_index, field) = match schema.column_with_name(s) {
```
Can we do the filtering perhaps inside the stats generation function? I think that will simplify it and would require the check only there

omkar-foss replied:
Sure, can you share the name of the function you're suggesting? I'll have a look.

omkar-foss commented on crates/core/src/writer/record_batch.rs:
```diff
@@ -1080,6 +1118,54 @@ mod tests {
     }
 
+    #[tokio::test]
+    async fn test_write_dssc_with_binary_col() {
```
@ion-elgreco this is the test associated with noop for binary cols while writing stats to json.

omkar-foss commented:
@ion-elgreco I've made the necessary changes to this PR as per our discussion. Please have a look when you get a chance, let me know any further changes required. Thanks! :)

ion-elgreco reviewed (approved):
Thanks for sticking with it 😄

omkar-foss commented:
@ion-elgreco looks like we're getting the same intermittent `BufError` failures again ([here](https://github.com/delta-io/delta-rs/actions/runs/13775527214/job/38540888622?pr=3146#step:5:116)), can you restart the failed ones?

omkar-foss commented:
The `BufError` [here](https://github.com/delta-io/delta-rs/actions/runs/13775527214/job/38540888622?pr=3146#step:5:116) seems related to this uv issue: https://github.com/astral-sh/uv/issues/8692

Probably we need to update the uv version in GitHub CI checks.

omkar-foss commented:
@ion-elgreco the [PyPI incident](https://status.python.org/incidents/2zf08fkp0nd1) shows as resolved now, shall we attempt to the re-run failing CI checks and try it out?
